middleware example for analytics (WIP) (#180)
vercel/examples: edge-functions/serverside-analytics/README.md

> # Serverside Analytics example

Reviewer: Adding "example" for titles is usually not required.
```suggestion
# Serverside Analytics
```

Author: Done 💯

> This example shows how to implement serverside analytics inside [Middleware Functions](https://nextjs.org/docs/middleware).

Reviewer: ```suggestion
This example shows how to use [Segment](https://segment.com) to implement server-side analytics at the edge with [Next.js Middleware](https://nextjs.org/docs/middleware).

[Segment](https://segment.com) allows users to collect data from multiple sources and push them to multiple destination. We are using it in this example for its flexibility.
```

Author: It's so edgy being at the edge 🤓. Done!

## Demo

> https://edge-serverside-analytics.vercel.app/

Reviewer: To make it match with other examples, the name should include `edge-functions`
```suggestion
https://edge-functions-serverside-analytics.vercel.app/
```

## How to Use

You can choose from one of the following two methods to use this repository:

### One-Click Deploy

Deploy the example using [Vercel](https://vercel.com?utm_source=github&utm_medium=readme&utm_campaign=next-example):

> [![Deploy with Vercel](https://vercel.com/button)](https://vercel.com/new/git/external?repository-url=https://github.com/vercel/examples/tree/main/edge-functions/serverside-analytics&project-name=serverside-analytics&repository-name=serverside-analytics)

Reviewer: We need to start adding the required env variables in the button, you can use the deploy button docs to create the link: https://vercel.com/docs/deploy-button

It needs 3 more query params, added by filling these fields:

<img width="744" alt="image" src="https://user-images.githubusercontent.com/4278345/157536820-5df80885-e51f-4add-80c8-2d69913a8975.png">

For the URL, add a link to the readme with a hash (`#`) pointing to the "Analytics Setup" title.

Author: Wow! Done!

### Clone and Deploy

Execute [`create-next-app`](https://github.com/vercel/next.js/tree/canary/packages/create-next-app) with [npm](https://docs.npmjs.com/cli/init) or [Yarn](https://yarnpkg.com/lang/en/docs/cli/create/) to bootstrap the example:

```bash
npx create-next-app --example https://github.com/vercel/examples/tree/main/edge-functions/serverside-analytics
# or
yarn create next-app --example https://github.com/vercel/examples/tree/main/edge-functions/serverside-analytics

```

> ### Analytics setup

Reviewer: ```suggestion
### Setting up Segment
```

> 
> [Segment](https://segment.com) is an analytics framework. It allows users to collect data from multiple sources and push them to multiple destination. We are using it in this example for it's flexibility.

Reviewer: ```suggestion
```

1.Register and create a workspace at [segment.com](https://segment.com)
2.Add a [javascript source](https://segment.com/docs/connections/sources/catalog/libraries/website/javascript/) and grab the `writeKey`
3.Add the write key in your `.env` file.

Next, run Next.js in development mode:

```bash
npm install
npm run dev

# or

yarn
yarn dev
```

Deploy it to the cloud with [Vercel](https://vercel.com/new?utm_source=github&utm_medium=readme&utm_campaign=edge-middleware-eap) ([Documentation](https://nextjs.org/docs/deployment)).

> **Note: Supabase was used to showcase page views analytics. You should research an appropriate database based on the volume and format of your Data.**

Reviewer: Why can't supabase be removed?

Author: Left over 🧹 . Done!

# Serverside Analytics example

> This example shows how to implement analytics inside [Middleware Functions](https://nextjs.org/docs/middleware).

Author: changed to segment, and added an explanation.

## Demo

> https://serverside-analytics.vercel.app

Reviewer 2: remember to prefix it with `edge-`

Author: Good point!

### Clone and Deploy

Execute [`create-next-app`](https://github.com/vercel/next.js/tree/canary/packages/create-next-app) with [npm](https://docs.npmjs.com/cli/init) or [Yarn](https://yarnpkg.com/lang/en/docs/cli/create/) to bootstrap the example:

```bash
npx create-next-app --example https://github.com/vercel/examples/tree/main/edge-functions/serverside-analytics
# or
yarn create next-app --example https://github.com/vercel/examples/tree/main/edge-functions/serverside-analytics
```

Next, run Next.js in development mode:

```bash
npm install
npm run dev

# or

yarn
yarn dev
```

Deploy it to the cloud with [Vercel](https://vercel.com/new?utm_source=github&utm_medium=readme&utm_campaign=edge-middleware-eap) ([Documentation](https://nextjs.org/docs/deployment)).

### Setting up:

1. Create a project on [Supabase](https://supabase.com/)
1. Get the API key and database URl and add them to `.env` and in your Vercel Project
1. Create a Table in Supabase containing the following data.

```
create table analytics (
  id bigint not null primary key,
  created_at timestamp default now(),
  slug character
);
```

> **Note: Supabase was used to showcase page views analytics. You should research an appropriate database based on the volume and format of your Data.**

Author: Done! also removed the supabase references.

### One-Click Deploy

Deploy the example using [Vercel](https://vercel.com?utm_source=github&utm_medium=readme&utm_campaign=next-example):

[![Deploy with Vercel](https://vercel.com/button)](https://vercel.com/new/git/external?repository-url=https://github.com/vercel/examples/tree/main/edge-functions/serverside-analytics&project-name=serverside-analytics&repository-name=serverside-analytics&env=SEGMENT_WRITE_KEY&envDescription=A%20Segment%20write%20key&envLink=https%3A%2F%2Fgithub.com%2Fvercel%2Fexamples%2Ftree%2Ffeat%2Fserverside-analytics%2Fedge-functions%2Fserverside-analytics%23setting-up-segment)

### Clone and Deploy

Execute [`create-next-app`](https://github.com/vercel/next.js/tree/canary/packages/create-next-app) with [npm](https://docs.npmjs.com/cli/init) or [Yarn](https://yarnpkg.com/lang/en/docs/cli/create/) to bootstrap the example:

```bash
npx create-next-app --example https://github.com/vercel/examples/tree/main/edge-functions/serverside-analytics
# or
yarn create next-app --example https://github.com/vercel/examples/tree/main/edge-functions/serverside-analytics

```

### Setting up Segment

1. Register and create a workspace at [segment.com](https://segment.com)

> 2. Add a [javascript source](https://segment.com/docs/connections/sources/catalog/libraries/website/javascript/) and grab the `writeKey`

Reviewer: @gbibeaul I did not understand what todo after clicking that link, let's add more detailed instructions in this step.
